mrmustard/lab_dev/sampler.py

```diff
@@ -0,0 +1,190 @@
+
+    @property
+    def povms(self) -> CircuitComponent | Sequence[CircuitComponent]:
+        r"""
+        The POVMs of this sampler.
+        """
+        return self._povms
+
+    @property
+    def meas_outcomes(self) -> Sequence[Any]:
+        r"""
+        The measurement outcomes of this sampler.
+        """
+        return self._meas_outcomes
+
+    @abstractmethod
+    def probabilities(self, state: State, atol: float = 1e-4) -> Sequence[float]:
+        r"""
+        Returns the probability distribution of a state w.r.t. measurement outcomes.
+
+        Args:
+            state: The state to generate the probability distribution of.
+            atol: The absolute tolerance used for validating the computed probability
+                distribution.
+        """
+
+    def sample(self, state: State, n_samples: int = 1000, seed: int | None = None) -> np.ndarray:
+        r"""
+        Returns an array of samples given a state.
+
+        Args:
+            state: The state to sample.
+            n_samples: The number of samples to generate.
+            seed: An optional seed for random sampling.
+        """
+        initial_mode = state.modes[0]
+        initial_samples = self.sample_prob_dist(state[initial_mode], n_samples, seed)
+
+        if len(state.modes) == 1:
+            return initial_samples
+
+        unique_samples, counts = np.unique(initial_samples, return_counts=True)
+        ret = []
+        for unique_sample, counts in zip(unique_samples, counts):
+            meas_op = self.povms[self.meas_outcomes.index(unique_sample)].on([initial_mode])
+            reduced_state = (state >> meas_op.dual).normalize()
+            samples = self.sample(reduced_state, counts)
+            for sample in samples:
+                ret.append(np.append([unique_sample], sample))
+        return np.array(ret)
+
+    def sample_prob_dist(
+        self, state: State, n_samples: int = 1000, seed: int | None = None
+    ) -> np.ndarray:
+        r"""
+        Samples a a state by computing the probability distribution.
+
+        Args:
+            state: The state to sample.
+            n_samples: The number of samples to generate.
+            seed: An optional seed for random sampling.
+        """
+        rng = np.random.default_rng(seed) if seed else settings.rng
+        return rng.choice(
+            a=list(product(self.meas_outcomes, repeat=len(state.modes))),
+            p=self.probabilities(state),
+            size=n_samples,
+        )
+
+    def _validate_probs(self, probs: Sequence[float], dx: float, atol: float) -> Sequence[float]:
+        r"""
+        Validates that the given probability distribution sums to `1.0` within some
+        tolerance and returns a renormalized probability distribution to account for
+        small numerical errors.
+
+        Args:
+            probs: The probability distribution to validate.
+            dx: The uniform differential for the probability distribution.
```

Comment:
I think this should be part of the probabilities method. This method should only accept probabilities, i.e. after the probability density has been multiplied by the differential element

Reply:
```suggestion
```